Move McpServer project template into its own package and add local vs. remote option

jeffhandley opened the pull request:
This is a draft, early feedback PR for introducing a `Microsoft.McpServer.ProjectTemplates` package for the `mcpserver` project template, bumping the template to use the latest version of ModelContextProtocol, and adding the a `--transport <local|remote>` option to the template.

```txt
Usage:
  dotnet new mcpserver [options] [template options]

Options:
  -n, --name <name>       The name for the output being created. If no name is specified, the name of the output directory is used.
  -o, --output <output>   Location to place the generated output.
  --dry-run               Displays a summary of what would happen if the given command line were run if it would result in a template creation. [default: False]
  --force                 Forces content to be generated even if it would change existing files. [default: False]
  --no-update-check       Disables checking for the template package updates when instantiating a template. [default: False]
  --project <project>     The project that should be used for context evaluation.
  -lang, --language <C#>  Specifies the template language to instantiate.
  --type <project>        Specifies the template type to instantiate.

Template options:
  -t, --transport <local|remote>           Whether to create a 'local' (stdio transport) or 'remote' (http transport) MCP server
                                           Type: choice
                                             local   A console application will be created to use the stdio transport as a local MCP server
                                             remote  An ASP.NET Core web application will be created to use the http transport as a remote MCP server
                                           Default: local
  --aot                                    Whether to enable the MCP server for publishing as a native AOT application.
                                           Type: bool
                                           Default: false
  --self-contained                         Whether to enable the MCP server for publishing as a self-contained application.
                                           Type: bool
                                           Default: true
  -f, --framework <net10.0|net8.0|net9.0>  The target framework for the project.
                                           Type: choice
                                             net10.0  .NET 10
                                             net9.0   .NET 9
                                             net8.0   .NET 8
                                           Default: net10.0
```

jeffhandley committed:
Move the mcpserver template into a new Microsoft.McpServer.ProjectTemplates package

jeffhandley committed:
Update ModelContextProtocol to 0.5.0-preview.1

jeffhandley committed:
Add local/remote transport type option to McpServer template

jeffhandley commented on src/ProjectTemplates/Microsoft.Extensions.AI.Templates/templates/McpServer-CSharp/.template.config/dotnetcli.host.json:
```diff
@@ -1,11 +1,6 @@
   "$schema": "https://json.schemastore.org/dotnetcli.host",
   "symbolInfo": {
-    "TargetFrameworkOverride": {
-      "isHidden": "true",
-      "longName": "target-framework-override",
-      "shortName": ""
-    },
```
This option was part of the first version of the template, with it coming from dotnet/templating sample code, but it's not in use, it was hidden, and it's undocumented, so it's safe to remove.

jeffhandley commented on src/ProjectTemplates/Microsoft.McpServer.ProjectTemplates/templates/McpServer-CSharp/remote/McpServer-CSharp.csproj-in:
```diff
@@ -0,0 +1,52 @@
+  <PropertyGroup>
+    <TargetFramework>net10.0</TargetFramework>
+    <TargetFramework Condition="'$(TargetFrameworkOverride)' != ''">TargetFrameworkOverride</TargetFramework>
+    <!--#if (SelfContained || NativeAot)-->
```
I wasn't certain that we would want to offer self-contained and native-aot on the remote server template. Let me know if you think we should remove that for now. We can make those options conditional to only be applicable when the local transport is used.

leslierichardson95 replied:
Why would we have these options available only for the local transport option?  Would these two options not be beneficial for remote servers?  Otherwise, I think I'd be ok with having these two options, maybe have them off default?

jeffhandley replied:
Having self-contained and aot enabled easily for a local MCP server is more important since it will be run on the user's machine vs. on your own web server. Enabling those options eliminates the need of having the shared framework installed on the user's machine. This can still apply to web apps but isn't as crucial. Having the defaults vary by `--transport` option could be tricky to achieve and/or confusing though.

It's worth noting though that the `--transport` option does change the project type more substantially than template options generally do, switching between a Console app and a Web app. The other approach that was discussed months back was to have 2 different project tempaltes:

1. `mcpserver` which is what we have today, a local/stdio Console app
2. `mcpserver-remote` (or some other name: `mcpserver-web`, `mcpserver-http`, etc.)

Discussions favored having the option on a single template.

jeffhandley commented on src/ProjectTemplates/Microsoft.McpServer.ProjectTemplates/templates/McpServer-CSharp/.template.config/template.json:
```diff
@@ -17,14 +17,44 @@
+          "description": "An ASP.NET Core web application will be created to use the http transport as a remote MCP server"
+        }
+      ],
+      "defaultValue": "local"
```
This defines the "local" transport as the default, which is backwards-compatible with the previous version of the template. But that could be changed if desired.

jeffhandley commented on src/ProjectTemplates/Microsoft.McpServer.ProjectTemplates/templates/McpServer-CSharp/remote/Program.cs:
```diff
@@ -0,0 +1,12 @@
+
+var app = builder.Build();
+app.MapMcp();
+
```
Note that I've left `UseHttpsRedirection()` out of here because of the issues with dev certificates in VS Code.

jeffhandley commented on src/ProjectTemplates/Microsoft.McpServer.ProjectTemplates/templates/McpServer-CSharp/remote/README.md:
````diff
@@ -0,0 +1,58 @@
+}
+```
+
+## Testing the MCP Server
+
+Once configured, you can ask Copilot Chat for a random number, for example, `Give me 3 random numbers`. It should prompt you to use the `get_random_number` tool on the `McpServer-CSharp` MCP server and show you the results.
````
If you have further ideas on content to add here, please don't hesitate to suggest it.

leslierichardson95 replied:
I added a comment in the Readme.md file that we should include a blurb about creating an mcp.json file to test the server

jeffhandley commented on src/ProjectTemplates/Microsoft.McpServer.ProjectTemplates/templates/McpServer-CSharp/remote/README.md:
```diff
@@ -0,0 +1,58 @@
+- [Official Documentation](https://modelcontextprotocol.io/)
+- [Protocol Specification](https://spec.modelcontextprotocol.io/)
+- [GitHub Organization](https://github.com/modelcontextprotocol)
+- [MCP C# SDK](https://modelcontextprotocol.github.io/csharp-sdk)
```
Adding a link to the C# SDK docs in both flavors of the README

jeffhandley commented on src/ProjectTemplates/Microsoft.McpServer.ProjectTemplates/templates/McpServer-CSharp/remote/README.md:
```diff
@@ -0,0 +1,58 @@
+
+Refer to the VS Code or Visual Studio documentation for more information on configuring and using MCP servers:
+
+- [Use MCP servers in VS Code (Preview)](https://code.visualstudio.com/docs/copilot/chat/mcp-servers)
```
I'm tempted to add a Known Issues section that can point to the VS Code issue for https. For the AIChatWeb template, we've noted known issues in that README as well and it's helped customers (including ourselves).

leslierichardson95 replied:
Makes sense to add "Known Issues" to me!

jeffhandley commented on src/ProjectTemplates/Microsoft.McpServer.ProjectTemplates/templates/McpServer-CSharp/remote/README.md:
```diff
@@ -0,0 +1,58 @@
+# MCP Server
+
+This README was created using the C# MCP server project template.
+It demonstrates how you can easily create an MCP server using C# and run it as an ASP.NET Core web application.
```
I removed all NuGet publishing related info from the remote version of the README and project file. Confirming: that was the expectation, correct?

leslierichardson95 replied:
I'm confused here.  Does it not make sense to publish a NuGet package when http is being used?

jeffhandley replied:
That is my understanding. [MCP servers in NuGet packages | Microsoft Learn](https://learn.microsoft.com/en-us/nuget/concepts/nuget-mcp#applicable-scenarios) cites:

> Consider the following criteria to determine whether shipping your MCP server as a NuGet package makes sense:
> 
> ✅ You want your MCP server to run locally on the user's system (i.e., in the same context as the MCP client).
> - Local MCP servers, such as those shipped in NuGet packages, run in the same context as the MCP client and communicate with the MCP client via standard IO (stdio) transport. The MCP client is responsible for launching the local MCP server process.

A remote MCP server would be deployed as a wep app. If you could help verify this with Joel, Maria, or others though I'd appreciate it.

jeffhandley commented on src/ProjectTemplates/Microsoft.McpServer.ProjectTemplates/templates/McpServer-CSharp/remote/README.md:
```diff
@@ -0,0 +1,58 @@
+  "servers": {
+    "McpServer-CSharp": {
+      "type": "http",
+      "url": "http://localhost:9996"
```
This uses the root URL; confirming that's the preferred address to connect to.

leslierichardson95 commented on src/ProjectTemplates/Microsoft.McpServer.ProjectTemplates/templates/McpServer-CSharp/remote/README.md:
```diff
@@ -0,0 +1,58 @@
+
+## Developing locally
+
+To test this MCP server from source code (locally), you can configure your IDE to connect to the server using localhost.
```
Maybe we should add a blurb here about creating an mcp.json file with the below information to be more specific?

jeffhandley replied:
Sure thing. Could you provide that content for me please? A snippet/suggestion here would be great.

jeffhandley replied:
We are moving the "Refer to the VS Code or Visual Studio documentation..." blurb up from the bottom of the README into this section, allowing their respective docs to describe it. VS Code uses `mcp.json` and VS uses `.mcp.json`. This allows us to avoid duplicating their docs into the template.

jeffhandley commented on src/ProjectTemplates/Microsoft.McpServer.ProjectTemplates/templates/McpServer-CSharp/remote/McpServer-CSharp.csproj-in:
```diff
@@ -0,0 +1,36 @@
+
+  <PropertyGroup>
+    <TargetFramework>net10.0</TargetFramework>
+    <TargetFramework Condition="'$(TargetFrameworkOverride)' != ''">TargetFrameworkOverride</TargetFramework>
```
I forgot to remove the `TargetFrameworkOverride` content from the remote flavor.

```suggestion
    <TargetFramework>net10.0</TargetFramework>
```

jeffhandley committed:
Remove TargetFrameworkOverride content from remote flavor

jeffhandley commented on test/ProjectTemplates/Microsoft.McpServer.ProjectTemplates.IntegrationTests/Snapshots/mcpserver/mcpserver.remote.verified/mcpserver/mcpserver.csproj:
```diff
@@ -0,0 +1,24 @@
+
+  <PropertyGroup>
+    <TargetFramework>net10.0</TargetFramework>
+    <TargetFramework Condition="'$(TargetFrameworkOverride)' != ''">TargetFrameworkOverride</TargetFramework>
```
```suggestion
    <TargetFramework>net10.0</TargetFramework>
```

jeffhandley committed:
Update snapshot to remove TargetFrameworkOverride

jeffhandley committed:
Address feedback on the mcpserver remote template

1. Move the VS/VS Code local config documentation blurb up in the README
2. Add a Known Issues section to the README
3. When using VS (not CLI or VS Code), use HTTPS instead of HTTP
